executor/cook/progress.py

```diff
@@ -0,0 +1,224 @@
+import cook
+
+
+class ProgressUpdater(object):
+    """This class is responsible for sending progress updates to the scheduler.
+    It throttles the rate at which progress updates are sent.
+    """
+
+    def __init__(self, poll_interval_ms, send_message_fn):
+        """
+        poll_interval_ms: int
+            The interval after which to send a subsequent progress update.
+        send_message: function(driver, message, max_message_length)
+            The helper function used to send the message.
+        """
+        self.poll_interval_ms = poll_interval_ms
+        self.last_reported_time = None
+        self.last_progress_data = None
+        self.send_message = send_message_fn
+        self.lock = Lock()
+
+    def has_enough_time_elapsed_since_last_update(self):
+        """Returns true if enough time (based on poll_interval_ms) has elapsed since 
+        the last progress update (available in last_reported_time).
+        """
+        if self.last_reported_time is None:
+            return True
+        else:
+            current_time = time.time()
+            time_diff_ms = (current_time - self.last_reported_time) * 1000
+            return time_diff_ms >= self.poll_interval_ms
+
+    def send_progress_update(self, driver, task_id, max_message_length, progress_data, force_send=False):
+        """Sends a progress update if enough time has elapsed since the last progress update.
+        The force_send flag can be used to ignore the check for enough time having elapsed.
+        Using this method is thread-safe.
+        
+        Parameters
+        ----------
+        driver: MesosExecutorDriver
+            The mesos driver to use.
+        task_id: string
+            The task id.
+        max_message_length: int
+            The allowed max message length after encoding.
+        progress_data: map 
+            The progress data to send.
+        force_send: boolean, optional
+            Defaults to false.
+            
+        Returns
+        -------
+        Nothing 
+        """
+        with self.lock:
+            if progress_data is not None and self.last_progress_data != progress_data:
+                if force_send or self.has_enough_time_elapsed_since_last_update():
+                    logging.info('Sending progress message {}'.format(progress_data))
+                    message_dict = dict(progress_data)
+                    message_dict['task-id'] = task_id
+                    progress_message = json.dumps(message_dict)
+
+                    if len(progress_message) > max_message_length and cook.PROGRESS_MESSAGE_KEY in message_dict:
+                        progress_str = message_dict[cook.PROGRESS_MESSAGE_KEY]
+                        num_extra_chars = len(progress_message) - max_message_length
+                        allowed_progress_message_length = max(len(progress_str) - num_extra_chars - 3, 0)
+                        new_progress_str = progress_str[:allowed_progress_message_length].strip() + '...'
```

Comment:
The strip is just so that the ellipsis appears after a non-whitespace character. Do you prefer it to be removed?